### DISCUSSION TITLE
Elastic Fracture of Functionally Graded Materials with Peridynamics Module

### DISCUSSION THREAD

**Discussion by ppandit95** [idaholab/moose#19588] [Q&A General] [answered]
Dear MOOSE Community

I am trying to simulate propagation of crack in a functionally graded material with a initial crack using NOSPD formulation and I need to apply stress as boundary condition on the top and bottom surface of the rectangular plate . So Any headers in this regard will be helpful.

Many Thanks
Pushkar

**Comment by GiudGiud** [comment]
Hello

The NOSPD models have theoretical documentation here:
https://mooseframework.inl.gov/modules/peridynamics/PeridynamicModels.html

And source/input documentation (how to choose between each model) here:
https://mooseframework.inl.gov/syntax/Modules/Peridynamics/Mechanics/Master/index.html

Guillaume

**Reply by ppandit95** [reply]
thanks @GiudGiud for directing to the relevent documentation for NOSPD model but I didn't see any documentation for Boundary Conditions for peridynamics module and so far in most of the tests DirichletBC has been used.So I was wondering is it the only BC that is applicable or can we apply NeumannBC and Robin BC as well

**Reply by jiangwen84** [reply]
@hchen139  I remembered you applied NeumannBC before. Need an additional layer? Could you comment here?

**Reply by hchen139** [reply]
You can first construct a sideset based on the FE sideset by setting the "construct_pd_sidesets = true" in your mesh block, and then apply the Neumann BC in the same way as in FE using the constructed sideset.

**Reply by ppandit95** [reply]
ohh... thanks @hchen139 for elaborating on the implementation of NeumannBCs

**Reply by ppandit95** [reply]
thanks @jiangwen84 and @GiudGiud for looking into this discussion thread and directing to relevant sources

**Comment by ppandit95** [comment]
As a first step to this simulation,I tried to simply extend the 2D stress based failure test by extending the end time up to which the simulation runs but encountered an error which says that "the shape tensor is singular" even though I have added SingularShapeTensorEliminatorUserObjectPD".Any header in this regard will be helpful as I am looking to reproduce a simulation of crack propagation in NOSPD material

**Comment by hchen139** [comment, marked as answer]
You may consider change the applied boundary conditions. In the input file, you fixed both left and right edges, while pulling the top and bottom edges in the positive and negative y directions, respectively. The crack would hardly open under these boundary conditions.



Hailong Chen, Ph.D.
Assistant Professor
Department of Mechanical Engineering
College of Engineering
University of Kentucky
279 Ralph G. Anderson Building
Lexington, KY 40506-0503
megroup.engr.uky.edu<http://megroup.engr.uky.edu/>
t: (859)323-8114
f: (859)257-3304
e: ***@***.***
________________________________

**Reply by ppandit95** [reply]
thanks @hchen139 for your suggestions on above issue and I did manage to get the damage model to work by changing BCs as you suggested but I am currently facing a error of singular shape tensor despite of adding shape tensor eliminator user object.

**Reply by Yunlong010413** [reply]
Hello, I'm getting the same error as you in my current simulation, how did you solve it then? Thank you! @ppandit95

**Reply by ppandit95** [reply]
Hi @Yunlong010413 ,although I am not completely sure this would resolve the issue but you can try reducing the time step

**Reply by GiudGiud** [reply]
@Yunlong010413 please open another post for this `shape tensor is singular` error

**Reply by hchen139** [reply]
@Yunlong010413 The reason for having a singular shape tensor in the simulation process is that there are not sufficient number of unbroken bonds for a particle to construct its shape tensor due to the bond breakage in the fracture modeling process. There is a way to avoid this by fixing the particle and assigning unit shape tensor, which is already implemented in the peridynamics module. Further, the shape tensor eliminator user object can be used to break all the bonds of a particle whose shape tensor is singular, since the particle is fixed anyway.

**Reply by Yunlong010413** [reply]
@ppandit95 Hi, thanks for the answers you gave  @GiudGiud .I started a post about this earlier,is  https://github.com/idaholab/moose/discussions/24004#discussioncomment-5562384
And @hchen139 Thank you for your answer, I have a general understanding of this problem, but I still don't know how to use the SingularShapeTensorEliminatorUserObjectPD function in the module, the specific situation is that after using it, it still appears: The following error occurred in the object "strain", of type "ComputePlaneSmallStrainNOSPD".
Singular shape tensor is detected in ComputeStrainBaseNOSPD! It may be because I just contacted MOOSE, many questions may seem very naive, thank you for your tolerance
